### PR TITLE
DFIQ Analyzer Implementation

### PR DESCRIPTION
**Pull request by jkppr**
This PR introduces a new DFIQ (Digital Forensics Investigative Questions) Analyzer framework to Timesketch, enabling the execution of analyzers based on investigative approaches defined in DFIQ objects.

**What existing problem does this PR solve?**

This PR addresses the need to integrate automated analysis workflows based on structured investigative methodologies like DFIQ.  It provides a way to automatically trigger relevant analyzers when a user adds DFIQ approaches to a sketch.

**What new feature is being introduced with this PR?**

* **DFIQ Analyzer Framework:** A new sub-framework for managing and executing DFIQ-aware analyzers.
* **Dynamic Analyzer Import:**  DFIQ analyzers are dynamically imported, providing flexibility and scalability.
* **Integration with Analyzer Framework:** Seamlessly integrates with the existing Timesketch analyzer framework, including status tracking and reporting.
* **API Endpoint Triggering:**  Triggers analysis execution when a DFIQ approach is added to a sketch.
* **Analysis-Approach Linking:**  Establishes a database relationship between `Analysis` and `InvestigativeQuestionApproach` models, enabling tracking of analyses related to specific approaches.

**Overview of changes to existing functions if required:**

* The `Analysis` model has been extended to include an optional relationship with `InvestigativeQuestionApproach`.
* The `tasks.py` file now supports DFIQ analyzer execution.
* A new `DFIQAnalyzerManager` class has been added to handle the execution of DFIQ analyzers.

**TODOs:**

- [x] Extend the analyzer framework for DFIQ
- [x] Trigger analyzer via DFIQ approaches
- [x] Trigger analyzer when new timelines are added
- [x] Link and fill InvestigativeQuestionConclusion by analyzer verdict.
- [x] Add unit tests
- [ ] Add one basic analyzer to ship with the PR
- [ ] Documentation updated.

This implementation provides a foundation for more advanced DFIQ-driven analysis workflows in Timesketch, allowing users to leverage structured methodologies for more efficient and guided investigations.

(The frontend part of this feature will be handled in a separate PR)

### CODE REVIEW

File: timesketch/lib/analyzers/dfiq_plugins/manager.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            try:
+                analyzer_manager.AnalysisManager.deregister_analyzer(name)
+                logger.info("Deregistered DFIQ analyzer: %s", name)
+            except KeyError as e:
+                logger.error(str(e))
+
+    # Dynamically load DFIQ Analyzers
+    for filename in os.listdir(DFIQ_ANALYZER_PATH):
+        if filename.endswith(".py") and not filename.startswith("__"):
+            module_name = filename[:-3]  # Remove .py extension
```
Comment (berggren) [inline timesketch/lib/analyzers/dfiq_plugins/manager.py:53]: Consider using path.splitext instead. It handles filenames with multiple dots etc and is a bit more robust.
```suggestion
            module_name = os.path.splitext(filename)[0]  # Remove .py extension
```

File: timesketch/models/sketch.py [hunk 2]
```diff
@@ -351,6 +351,10 @@ class Analysis(GenericAttributeMixin, LabelMixin, StatusMixin, CommentMixin, Bas
     sketch_id = Column(Integer, ForeignKey("sketch.id"))
     timeline_id = Column(Integer, ForeignKey("timeline.id"))
     searchindex_id = Column(Integer, ForeignKey("searchindex.id"))
+    approach_id = Column(Integer, ForeignKey("investigativequestionapproach.id"))
+    question_conclusion_id = Column(
```
Comment (berggren) [inline timesketch/models/sketch.py:355]: I know this is long and ugly, but to stay consistent in the file, please rename:
```suggestion
    investigativequestionconclusion_id = Column(
```

File: timesketch/models/sketch.py [hunk 3]
```diff
@@ -351,6 +351,10 @@ class Analysis(GenericAttributeMixin, LabelMixin, StatusMixin, CommentMixin, Bas
     sketch_id = Column(Integer, ForeignKey("sketch.id"))
     timeline_id = Column(Integer, ForeignKey("timeline.id"))
     searchindex_id = Column(Integer, ForeignKey("searchindex.id"))
+    approach_id = Column(Integer, ForeignKey("investigativequestionapproach.id"))
```
Comment (berggren) [inline timesketch/models/sketch.py:354]: I know this is long and ugly, but to stay consistent in the file, please rename:
```suggestion
    investigativequestionapproach_id = Column(Integer, ForeignKey("investigativequestionapproach.id"))
```

**Review by berggren — CHANGES_REQUESTED** [review]
First round. Looks good, some minor comments

File: timesketch/lib/analyzers/dfiq_plugins/__init__.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import timesketch.lib.analyzers.dfiq_plugins.manager
+
+logger = logging.getLogger("timesketch.lib.analyzers.dfiq_plugins")
+
+# Dynamically load DFIQ Analyzers
+DFIQ_ANALYZER_PATH = os.path.dirname(os.path.abspath(__file__))
+
+for filename in os.listdir(DFIQ_ANALYZER_PATH):
+    if filename.endswith(".py") and not filename.startswith("__"):
+        module_name = filename[:-3]  # Remove .py extension
```
[the remark this answers is not in the document]
Comment (jkppr) [reply timesketch/lib/analyzers/dfiq_plugins/__init__.py]: This moved to the manager.py, but I'll change it there.

File: timesketch/api/v1/resources/analysis.py [hunk 5]
```diff
@@ -188,22 +190,26 @@ def get(self, sketch_id):
...
-            if len(analyzer_class.get_kwargs()) > 0:
-                multi = True
             analyzers_detail.append(
                 {
                     "name": analyzer_name,
                     "display_name": analyzer_class.DISPLAY_NAME,
                     "description": analyzer_class.DESCRIPTION,
-                    "is_multi": multi,
+                    "is_multi": len(analyzer_class.get_kwargs()) > 0,
+                    "is_dfiq": hasattr(analyzer_class, "IS_DFIQ_ANALYZER")
```
[the remark this answers is not in the document]
Comment (jkppr) [reply timesketch/api/v1/resources/analysis.py:211]: I guess you are referring to `hasattr(analyzer_class, "IS_DFIQ_ANALYZER") and analyzer_class.IS_DFIQ_ANALYZER`?
All analyzers should have the attribute from the BaseAnalyzer class. But to catch situations I did not think of where an analyzer does not have this attribute I added the first condition. The second condition sets the value True for DFIQ analyzers (or false if no attribute exists or is a classic analyzer).
So following this logic, I think we need both?

File: timesketch/models/sketch.py (hunk 3 above)
Comment (jkppr) [reply timesketch/models/sketch.py:354]: We are using `approach_id` already in the models: https://github.com/google/timesketch/blob/master/timesketch/models/sketch.py#L464 

Still want it to have changed? I'm fine with both.

File: timesketch/api/v1/resources/analysis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (berggren) [reply timesketch/api/v1/resources/analysis.py:211]: Ack, makes sense

File: timesketch/models/sketch.py (hunk 3 above)
Comment (berggren) [reply timesketch/models/sketch.py:354]: Let's keep the short version 👍🏼

**Review by berggren — APPROVED** [review]
LGTM

**Comment by jkppr** [comment]
Going to merge the current state and track next steps in #3198

**Pull request merged**